File: src/main/java/de/tectoast/emolga/commands/draft/RandomPickCommand.java
package de.tectoast.emolga.commands.draft;

import de.tectoast.emolga.commands.Command;
import de.tectoast.emolga.commands.CommandCategory;
import de.tectoast.emolga.commands.GuildCommandEvent;
import de.tectoast.emolga.utils.draft.Draft;
import de.tectoast.emolga.utils.draft.Tierlist;
import net.dv8tion.jda.api.entities.Member;
import net.dv8tion.jda.api.entities.TextChannel;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.function.Predicate;

public class RandomPickCommand extends Command {
    public RandomPickCommand() {
        super("randompick", "Well... nen Random-Pick halt", CommandCategory.Draft);
        setArgumentTemplate(ArgumentManagerTemplate.builder()
                .add("tier", "Tier", "Das Tier, in dem gepickt werden soll", ArgumentManagerTemplate.Text.any())
                .addEngl("type", "Typ", "Der Typ, von dem random gepickt werden soll", Translation.Type.TYPE, true)
                .setExample("!randompick A")
                .build());
    }

    @Override
    public void process(GuildCommandEvent e) {
        Member memberr = e.getMember();
        long member = memberr.getIdLong();
        TextChannel tco = e.getChannel();
        Draft d = Draft.getDraftByMember(member, tco);
        if (d == null) {
            e.getChannel().sendMessage("Du Kek der Command funktioniert nur in einem Draft xD").queue();
            return;
        }
        Tierlist tierlist = d.getTierlist();
        ArgumentManager args = e.getArguments();
        String tier = tierlist.order.stream().filter(s -> args.getText("tier").equalsIgnoreCase(s)).findFirst().orElse("");
        if (tier.isEmpty()) {
            tco.sendMessage("Das ist kein Tier!").queue();
            return;
        }
        if (!d.tc.getId().equals(tco.getId())) return;
        if (d.isNotCurrent(member)) {
            tco.sendMessage(d.getMention(member) + " Du bist nicht dran!").queue();
            return;
        }
        long mem = d.current;
        List<String> list = new ArrayList<>(tierlist.tierlist.get(tier));
        Collections.shuffle(list);
        Predicate<String> typecheck;
        if (args.has("type")) {
            Translation type = args.getTranslation("type");
            typecheck = str -> getDataJSON().getJSONObject(getSDName(str)).getJSONArray("types").toList().contains(type.getTranslation());
        } else {
            typecheck = str -> true;
        }
        PickCommand.exec(tco, "!pick " + list.stream().filter(str -> !d.isPicked(str) && !d.hasInAnotherForm(mem, str) && (!d.hasMega(mem) || !str.startsWith("M-")) && typecheck.test(str)).map(String::trim).findFirst().orElse(""), memberr, true);
    }
}

File: src/main/kotlin/de/tectoast/emolga/commands/draft/RandomPickCommand.kt
package de.tectoast.emolga.commands.draft

import de.tectoast.emolga.commands.*
import de.tectoast.emolga.database.exposed.NameConventionsDB
import de.tectoast.emolga.utils.Constants
import de.tectoast.emolga.utils.draft.isEnglish
import de.tectoast.emolga.utils.json.db
import de.tectoast.emolga.utils.json.emolga.draft.League
import de.tectoast.emolga.utils.json.get
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import java.util.concurrent.ConcurrentHashMap

object RandomPickCommand : Command("randompick", "Well... nen Random-Pick halt", CommandCategory.Draft) {
    init {
        argumentTemplate = ArgumentManagerTemplate.builder()
            .add("tier", "Tier", "Das Tier, in dem gepickt werden soll", ArgumentManagerTemplate.Text.any())
            .addEngl("type", "Typ", "Der Typ, von dem random gepickt werden soll", Translation.Type.TYPE, true)
            .setExample("/randompick A").build()
        slash(true, Constants.G.ASL, Constants.G.FLP, Constants.G.WFS)
    }


    private val tierRestrictions = mapOf(
        Constants.G.ASL to setOf("D")
    )
    private val locks = ConcurrentHashMap<String, Mutex>()


    override suspend fun process(e: GuildCommandEvent) {
        val d =
            League.byCommand(e)?.first ?: return e.reply(
                "Es läuft zurzeit kein Draft in diesem Channel!",
                ephemeral = true
            )
        val mutex = locks.getOrPut(d.leaguename) { Mutex() }
        mutex.withLock {
            val tierlist = d.tierlist
            val args = e.arguments
            val gid = e.guild.idLong
            val tier = (tierlist.order.firstOrNull { args.getText("tier").equals(it, ignoreCase = true) }
                ?: return e.reply("Das ist kein Tier!"))
                .takeIf {
                    tierRestrictions[gid]?.run { isEmpty() || contains(it) } != false
                }
                ?: return e.reply("In dieser Liga darf nur in folgenden Tiers gerandompickt werden: ${tierRestrictions[gid]?.joinToString()}")

            val list = tierlist.getByTier(tier)!!.shuffled()
            val typecheck: (suspend (String) -> Boolean) = if (args.has("type")) {
                val type = args.getTranslation("type");
                { type.translation in db.pokedex.get(it.toSDName())!!.types }
            } else { _ -> true }
            e.arguments.map["pokemon"] = (list.firstNotNullOfOrNull { str: String ->
                val draftName = NameConventionsDB.getDiscordTranslation(
                    str, d.guild, tierlist.isEnglish
                )!!
                draftName.takeIf {
                    !d.isPicked(
                        draftName.official, tier
                    ) && typecheck(
                        NameConventionsDB.getDiscordTranslation(
                            str, d.guild, true
                        )!!.official
                    )
                }
            } ?: return e.reply("In diesem Tier gibt es kein Pokemon mit dem angegebenen Typen mehr!"))

            PickCommand.exec(e, true)
        }
    }
}
